src/ORM/FieldType/DBField.php

```diff
@@ -542,7 +542,11 @@ public function saveInto($dataObject)
                 "DBField::saveInto() Called on a nameless '" . static::class . "' object"
             );
         }
-        $dataObject->setField($fieldName, $this->value);
+        if ($this->value instanceof self) {
+            $this->value->saveInto($dataObject);
+        } else {
+            $dataObject->__set($fieldName, $this->value);
+        }
```

Comment:
```suggestion
        $dataObject->$fieldName = $this->value;
```
We don't need `$this->value->saveInto($dataObject)` as it will be called [here](https://github.com/silverstripe/silverstripe-framework/pull/10614/files#diff-870d29e65995b805fda5b1fe74f788cf44cd8746b7f6edf224d5ea76b5159b93L2838) if $val is a DBField.

We should be using the arrow notation that calls __set() rather than calling __set() directly because there's some infinite recursion protection which I think happens at the PHP level, and also because calling __set() directly just looks weird.

Reply:
> We don't need $this->value->saveInto($dataObject) as it will be called [here](https://github.com/silverstripe/silverstripe-framework/pull/10614/files#diff-870d29e65995b805fda5b1fe74f788cf44cd8746b7f6edf224d5ea76b5159b93L2838) if $val is a DBField.

This shortcuts passed that, avoiding repeated calls through this chain when it's not needed. Imagine the (admittedly unlikely) scenario where a `DBField` instance has another `DBField` instance as its value, which has _another_ `DBField` instance as its value, etc. What you're proposing will go through the `saveInto() => __set() => setField() => saveInto()` loop for every instance. The way I've got it, all of that is resolved in `saveInto()` so that only the raw value gets passed through that series.

The `if ($val instanceof DBField)` check is necessary for when we do `$obj->MyField = new DBField()` i.e. set the dbfield instance to the value _as though the value is a property_. Imagine again that this new `DBField` we're setting as the value has a series of nested `DBField` values - again, your suggestion results in it going through the loop a bunch of times. With the way I've got it, we just go `__set() => setField() => saveInto()` at which point we resolve down to the raw value and only go down to `__set()` one last time.

Reply:
> We should be using the arrow notation that calls __set() rather than calling __set() directly

What do you mean "the arrow notation"? Do you mean just pretend we're setting a property? If we do that, the tests fail.

> because there's some infinite recursion protection only when using arrow notation which I think happens at the PHP level

The _only_ way we'll run into that scenario is if a `DBField` has _itself_ in the value chain, which is _extremely_ unlikely. But if you like I can write some logic to detect that in `saveInto` which wouldn't be difficult to do.

> because calling __set() directly just looks weird

That's not a good reason to not do the thing that preserves the expected behaviour. This is the only way to get the tests I've written passing, from what I can tell.

Reply:
>> What do you mean "the arrow notation"? Do you mean just pretend we're setting a property?

Yes, that will call __set(). We should be able safety assume the property doesn't exist on the DataObject since it's in the context of DBField->saveInto(), where $fieldName is DBField->name and Silverstripe ORM revolves around DataObject->FieldName magic.

>> which is extremely unlikely

I triggered it 2 different ways while investigating ways to simplify this PR. Chris's example code calls setField(), so does ViewableData::__set(), so it doesn't seem too hard, particularly when some related code gets modified in the future.

>> This shortcuts passed that, avoiding repeated calls through this chain when it's not needed

I'd rather take the small performance hit of repeated calls (we're talking microseconds) and ensure that we're calling all logic that should be called, rather than fragmenting the code.

Reply:
> Yes, that will call __set(). We should be able safety assume the property doesn't exist on the DataObject since it's in the context of DBField->saveInto(), where $fieldName is DBField->name and Silverstripe ORM revolves around DataObject->FieldName magic.

But it _won't_ call `__set()` if we're already coming from `__set()` which is the case when setting values as properties, which is the most common scenario. As I mentioned before, it will cause the tests to fail. I'm happy to write some quick anti-recursion logic to avoid the unlikely recursive scenario, but we can't just treat fields as properties from the dbfield because it _won't always call `__set()`_ and will result in failed tests.

> I triggered it 2 different ways while investigating ways to simplify this PR. Chris's example code calls setField(), so does ViewableData::__set(), so it doesn't seem too hard, particularly when some related code gets modified in the future.

Please provide a clear set of instructions for triggering the problematic behaviour, so that I can write a test against it and update the code to pass that test.

> I'd rather take the small performance hit of repeated calls (we're talking microseconds) and ensure that we're calling all logic that should be called, rather than fragmenting the code.

It's not just about performance. People creating their own custom `setMyField()` setter methods won't be expecting those to get hit multiple times each time a field is set. I've got it down to a _minimum_ of two times, which is already one two many but I think it's unavoidable having that second call while preserving otherwise expected behaviour. We should avoid repeated calls as it may cause unexpected side-effects in custom code which doesn't expect to be called multiple times per field set. "all logic that should be called" _is_  being called in this PR. We're getting _the `DBField` instance which has the raw value_ and then calling `saveInto()` on that field, and then going through _all_ of the `__set()` logic.

Reply:
I'll gladly add additional recursion protection if you want me to, as I've stated. I'll also gladly add tests against the scenario you've triggered 2 different ways if you tell me what those 2 different ways are, and then update the code to make those tests pass.
I won't change this PR to call `$obj->$field = $val` because I categorically think that's incorrect.

Reply:
The whole __set() / setField() / setFieldValue() / saveInto() thing is so convoluted :-/

I see your point about having to call __set(), because we need to bypass the recursion protection that the arrow comes with. 

> The if ($val instanceof DBField) check is necessary for when we do $obj->MyField = new DBField() i.e. set the dbfield instance to the value as though the value is a property. 

Is this actually a real-world thing?  Having to support the `$val instance of DBField` in DataObject::setField() is basically the root cause of the complexity here.

I did a search on installer for the following regexs
- `->[A-Za-z]+ = new [A-Za-z]+Field`
- `->[A-Za-z]+ = [A-Za-z]+Field::create\(`
In every instances it's just assignments to some a class property, not an assignment to a property on a dataobject

Reply:
> Is this actually a real-world thing? Having to support the `$val instance of DBField` in DataObject::setField() is basically the root cause of the complexity here.

If we were still pre-beta I'd say lets just get rid of that and see what breaks.... but I think it's too late to remove that. At best we could mark passing `DBField` in as property as a deprecated behaviour so it'll be easy to remove in 6, but I think it's risky to remove it now. It feels like the sort of thing that there's some magic functionality somewhere we're unaware of with bad test coverage that will break and we won't realise until someone complains.
That said, if you wanna get Max's approval to remove it (we'd need that since it'd be a breaking change post-beta) I'm not super opposed to the idea.

Reply:
Maybe we need an extra set of eyes across this to sanity check it? Or do you think it's okay now that you've understood where I was coming from by calling `__set()`? Or.... ??? Not sure what the next course of action is here?

Reply:
I've had look into the feasibility of removing `$val->saveInto($this)` within `DataObject::setField()` isn't viable.  The reason comes down to DBComposite fields where the following notation is supported, for instance for DBMoney

`$dataObject->MyDBField->methodOnDBField()`